modules/kernel-config.nix

```diff
@@ -83,6 +83,37 @@ in
         BLK_DEV_BSG = yes;
         DEVPTS_MULTIPLE_INSTANCES = whenOlder "4.7" yes;
       })
+      # Needed for waydroid
+      (helpers: with helpers; let
+        inherit (lib) mkMerge;
+        # TODO drop when we fix modular kernels
+        module = yes;
+      in {
+        ANDROID = whenAtLeast "3.19" yes;
+        ANDROID_BINDER_IPC = whenAtLeast "3.19" yes;
+        ANDROID_BINDERFS = whenAtLeast "5.0" yes;
+
+        # Needed for waydroid networking to function
+        NF_TABLES = whenAtLeast "3.13" yes;
+        NF_TABLES_IPV4 = mkMerge [ (whenBetween "3.13" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NF_TABLES_IPV6 = mkMerge [ (whenBetween "3.13" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NF_TABLES_INET = mkMerge [ (whenBetween "3.14" "4.17" module) (whenAtLeast "4.17" yes) ];
+        NFT_MASQ = whenAtLeast "3.18" module;
+        NFT_NAT = whenAtLeast "3.13" module;
+        IP_ADVANCED_ROUTER = whenAtLeast "2.5.45" yes;
+        IP_MULTIPLE_TABLES = whenAtLeast "2.5.45" yes;
+        IPV6_MULTIPLE_TABLES = whenAtLeast "2.6.19" yes;
+
+        # Needed for XfrmController
+        XFRM = whenAtLeast "2.6.0" yes;
+        XFRM_ALGO = whenAtLeast "3.5" module;
+        XFRM_USER = whenAtLeast "2.6.19" module;
+
+        # netd uses NFLOG
+        NETFILTER_NETLINK = whenAtLeast "2.6.25" yes;
+        NETFILTER_NETLINK_LOG = whenAtLeast "2.6.14" yes;
+        NETFILTER_XT_TARGET_NFLOG = whenAtLeast "2.6.20" module;
```

Comment:
We don't need to gate anything older than the oldest in-use kernel version, `3.9` here.

Even `3.9` is getting too old for some things that e.g. systemd requires.

Also noting, but nothing to do for you, that we'll see in the future if it becomes problematic to re-define some symbol names like the networking bits here. I guess that as long as it resolves to the same answer it's a non-issue.

Reply:
> We don't need to gate anything older than the oldest in-use kernel version, `3.4` here.
> 
> Even `3.4` is getting too old for some things that e.g. systemd requires.

Removed unnecessary gates.

> Also noting, but nothing to do for you, that we'll see in the future if it becomes problematic to re-define some symbol names like the networking bits here. I guess that as long as it resolves to the same answer it's a non-issue.

I'd definitely like to see this being possible such that we can fully define the requirements for a given concern without any dependency. This way regardless of what happens with the firewall etc. waydroid should always function correctly as it's fully specified outside of any context. Just less fragile that way.

I can see a problematic corner-case with this, though: one place defines a config item as a module, another built-in. It would probably be best to make sure we have an assertion/warning if that occurs. Not important currently as we don't have modules, but thinking ahead here.

Reply:
Yeah, it's more of a thought for *whenever that happens in the future*.

These configs aren't user-facing; they are for implementing a default kernel config. It is not expected that end-users will customize those, and if they will, that they know how to resolve the issues.